### PR TITLE
New Person Page (V2)

### PR DESCRIPTION
**Pull request by paolodamico**
## Changes

Implements the new person detail page (V2) from #2353 (everything behind feature flag `persons-2353`). It also allows editing properties inline.

In addition,
- Uses the existing `PropertiesTable` and deprecates `PersonTable`.
- Moves all relevant components to TS and Kea.

Next improvements (separate PR):
- Search/filter properties
- Create properties
- Vertical UI design for properties

Future improvements:
- Support changing date with a datetime selector.
- Allow editing nested properties (i.e. arrays and objects)
- Show properties in a JSON format to allow more complex editing ?

## Checklist

- [X] All querysets/queries filter by Organization, by Team, and by User
- [X] Django backend tests
- [X] Jest frontend tests
- [X] Cypress end-to-end tests

### CODE REVIEW

File: frontend/src/scenes/persons/MergePerson.tsx [hunk 1]
```diff
@@ -24,62 +25,48 @@ export function MergePersonButton({
...
+                    posthog.capture('merge person completed', { merge_count: selectedPeople.length })
+                    closeModal()
+                    onPersonChange(newPerson)
+                }
+            }}
+        >
+            Merge all properties and events of the selected users into <strong>{person.name}.</strong>
+            <br />
+            <br />
+            Note: If there is a conflict, the properties of <strong>{person.name}</strong> will take precedence.
```
Comment (macobo) [inline frontend/src/scenes/persons/MergePerson.tsx:47]: Might be not part of this change, but should we indicate this is not reversible?

File: frontend/src/scenes/persons/PersonProperty.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import { Tooltip } from 'antd'
+
+export function PersonProperty({
+    name,
+    value,
+}: {
+    name: string
+    value: string | number | boolean | null
+}): JSX.Element {
+    const type = useMemo(() => {
```
Comment (macobo) [inline frontend/src/scenes/persons/PersonProperty.tsx:14]: This seems a bit suss - don't we have that logic already in frontend/src/lib/components/PropertiesTable.tsx?

File: frontend/src/scenes/persons/PersonV2.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    const ids = (
+        <Menu>
+            {person?.distinct_ids.map((distinct_id: string) => {
+                return (
+                    <Menu.Item key={distinct_id}>
+                        <CopyToClipboardInline
+                            explicitValue={distinct_id}
+                            tooltipMessage=""
+                            iconStyle={{ color: 'var(--primary)' }}
```
Comment (macobo) [inline frontend/src/scenes/persons/PersonV2.tsx:35]: would it make sense to use classnames here? Including css vars in react like this might lead to errors when e.g. refactoring something

File: frontend/src/scenes/persons/PersonV2.tsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+function _PersonV2(): JSX.Element {
+    const [activeTab, setActiveTab] = useState('events')
+    const [mergeModalOpen, setMergeModalOpen] = useState(false)
+
+    const { person, personLoading, deletedPersonLoading } = useValues(personsLogic)
+    const { deletePerson, setPerson } = useActions(personsLogic)
+
+    const ids = (
+        <Menu>
+            {person?.distinct_ids.map((distinct_id: string) => {
```
Comment (macobo) [inline frontend/src/scenes/persons/PersonV2.tsx:29]: Nit: This can be simplified to `distinct_id: string) => (` and drop the explicit return below

File: frontend/src/scenes/persons/PersonV2.tsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                        <Tabs>
+                            <TabPane
+                                tab={<span data-attr="persons-properties-tab">Properties</span>}
+                                key="properties"
+                                disabled={personLoading}
+                            />
+                        </Tabs>
+                        {person && (
+                            <>
+                                {Object.keys(person.properties)
```
Comment (macobo) [inline frontend/src/scenes/persons/PersonV2.tsx:141]: This is duplicating frontend/src/lib/components/PropertyKeyInfo.js which we already duplicate badly in other places (see e.g. not being able to filter by props starting with $ in some modals).

We should consolidate this logic rather than pull this into more specific components.

File: frontend/src/scenes/persons/PersonsTableV2.tsx [hunk 6]
```diff
@@ -33,7 +32,7 @@ export function PersonsTable({
     cohort,
 }: PersonsTableType): JSX.Element {
     const linkToPerson = (person: PersonType): string => {
-        const backTo = cohort ? `#backTo=Back%20to%20Cohorts&backToURL=/cohorts/${cohort.id}` : ''
+        const backTo = cohort ? `#backTo=Cohorts&backToURL=/cohorts/${cohort.id}` : '#backTo=Persons&backToURL=/persons'
```
Comment (macobo) [inline frontend/src/scenes/persons/PersonsTableV2.tsx:35]: This seems sketch. Would `location.pathname` make more sense here?

File: frontend/src/scenes/persons/personsLogic.ts [hunk 7]
```diff
@@ -52,6 +60,34 @@ export const personsLogic = kea<personsLogicType<PersonPaginatedResponse>>({
                 },
             },
         ],
+        person: [
+            null as PersonType | null,
+            {
+                loadPerson: async (id) => {
```
Comment (macobo) [inline frontend/src/scenes/persons/personsLogic.ts:66]: Let's add an explicit type to `id` as well as for the return value here - otherwise ts thinks this is `any`

File: frontend/src/scenes/persons/personsLogic.ts [hunk 8]
```diff
@@ -52,6 +60,34 @@ export const personsLogic = kea<personsLogicType<PersonPaginatedResponse>>({
...
+                    const response = await api.get(`api/person/?distinct_id=${id}`)
+                    if (!response.results.length) {
+                        router.actions.push('/404')
+                    }
+                    return response.results[0]
+                },
+                setPerson: (person: PersonType) => {
```
Comment (macobo) [inline frontend/src/scenes/persons/personsLogic.ts:73]: Simplify: `setPerson: (person: PersonType) => person`

File: frontend/src/scenes/persons/personsLogic.ts [hunk 9]
```diff
@@ -52,6 +60,34 @@ export const personsLogic = kea<personsLogicType<PersonPaginatedResponse>>({
...
+        ],
+        deletedPerson: [
+            false,
+            {
+                deletePerson: async () => {
+                    if (!values.person) {
+                        return
+                    }
+                    await api.delete(`api/person/${values.person.id}`)
+                    return true
```
Comment (macobo) [inline frontend/src/scenes/persons/personsLogic.ts:87]: Is this read anywhere?

**Review by macobo — COMMENTED** [review]
Left some mainly code-level questions/suggestions. Flow works really well!

File: frontend/src/scenes/persons/MergePerson.tsx (hunk 1 above)
Comment (paolodamico) [reply frontend/src/scenes/persons/MergePerson.tsx:47]: Not part of the PR, but I completely agree and it's a very minor change, I've included it.

File: frontend/src/scenes/persons/PersonProperty.tsx (hunk 2 above)
Comment (paolodamico) [reply frontend/src/scenes/persons/PersonProperty.tsx:14]: I was thinking of introducing instead, but it might not make sense right now. I've kept the logic from PropertiesTable, and in a separate PR I'll update the UI (feature flagged) to the new layout.

File: frontend/src/scenes/persons/PersonV2.tsx (hunk 3 above)
Comment (paolodamico) [reply frontend/src/scenes/persons/PersonV2.tsx:35]: As we do this same thing in a couple of other components, I'll keep it for a separate small PR refactor using class names.

File: frontend/src/scenes/persons/PersonV2.tsx (hunk 5 above)
Comment (paolodamico) [reply frontend/src/scenes/persons/PersonV2.tsx:141]: I'm actually using `PropertyKeyInfo.js` inside the component, the extra logic is for sorting purposes only. However, you're 100% right, I will add this logic to the main `PropertiesTable` component so it's reused across components.

File: frontend/src/scenes/persons/personsLogic.ts (hunk 7 above)
Comment (paolodamico) [reply frontend/src/scenes/persons/personsLogic.ts:66]: I've added the return type to the function, though I'm not sure if we should do this. It was my understanding that the kea typegen magic handle this automatically based on the type set on line 64, @mariusandra ?

File: frontend/src/scenes/persons/personsLogic.ts (hunk 9 above)
Comment (paolodamico) [reply frontend/src/scenes/persons/personsLogic.ts:87]: No, but it seems that if we don't return a value the reducer raises an exception and the action lifecycle is not completed.

Comment (macobo) [reply frontend/src/scenes/persons/personsLogic.ts:87]: Aha. That's because this is marked as a loader - even though it isn't loading any data that's being used.

I think this should be a listener: https://kea.js.org/docs/guide/concepts#listeners

File: frontend/src/scenes/persons/personsLogic.ts [hunk 10]
```diff
@@ -52,6 +60,34 @@ export const personsLogic = kea<personsLogicType<PersonPaginatedResponse>>({
...
+            null as PersonType | null,
+            {
+                loadPerson: async (id: string): Promise<PersonType> => {
+                    const response = await api.get(`api/person/?distinct_id=${id}`)
+                    if (!response.results.length) {
+                        router.actions.push('/404')
+                    }
+                    return response.results[0]
+                },
+                setPerson: (person: PersonType): PersonType => {
```
Comment (macobo) [inline frontend/src/scenes/persons/personsLogic.ts:73]: This should be in the reducer as loaders should load data from API.

File: frontend/src/scenes/persons/personsLogic.ts (hunk 7 above)
Comment (macobo) [reply frontend/src/scenes/persons/personsLogic.ts]: Kea will automatically infer the types typescript is able to infer - in this case though it thinks it's dealing with `any` which is a unescapable black box.

File: frontend/src/scenes/persons/PersonV2.tsx (hunk 5 above)
Comment (macobo) [reply frontend/src/scenes/persons/PersonV2.tsx]: :heart:

**Comment by macobo** [comment]
Some minor code nits and this needs a rebase (to remove the commits from your earlier branch) - otherwise good to go! Good job.

File: frontend/src/scenes/persons/personsLogic.ts (hunk 10 above)
Comment (paolodamico) [reply frontend/src/scenes/persons/personsLogic.ts:81]: That doesn't work because `person` is registered in a loader, if I registered the `person` reducer to be able to respond to a `setPerson` action, it messes with the loader (because the loader automatically creates a reducer with the same name). The alternative is manually register the actions, reducers & listeners that the loader creates, but I don't think that's worth it.

**Comment by paolodamico** [comment]
All comments addressed and ready for a final review. There's still a few extra improvements I want to add but don't want to bloat the PR, will do them separately.

**Comment by macobo** [comment]
Not 100% confident on the loader stuff but everything else looks great. Merging :)
